### PR TITLE
Section 6 154-168

### PR DESCRIPTION
**Pull request by seanho96**
- Create and destroy user sessions
- Add authentication helper methods
- Declare controller methods as helper methods
- Restrict actions from UI
- Modify navigation based on helpers
- Restrict action from controller for articles and users
- Delete user

### CODE REVIEW

File: app/controllers/pages_controller.rb [hunk 1]
```diff
@@ -1,4 +1,6 @@
 class PagesController < ApplicationController
-  def home; end
+  def home
+    redirect_to articles_path if logged_in?
+  end
```
Comment (tlkiong) [inline app/controllers/pages_controller.rb:4]: Give a space in between 2 methods

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -29,15 +30,34 @@ def update
...
+      session[:user_id] = @user.id
       flash[:notice] = "Welcome to the Meta Blog #{@user.username}"
       redirect_to articles_path
     else
       render 'new'
     end
   end
 
+  def destroy
+    @user.destroy
```
Comment (tlkiong) [inline app/controllers/users_controller.rb]: Whats the difference between `.destroy` & `.delete`?

Comment (seanho96) [reply app/controllers/users_controller.rb]: `.destroy` can run callbacks on the model before running the `DELETE` query from the db while `.delete` doesn't. `Destroy` can be used if our models have association or relationships with other models, where we can also destroy associated data when calling destroy on an object with the help of helpers like `:dependent => :destroy_all` before it gets deleted.

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,2 +1,18 @@
...
+  def current_user
+    @current_user ||= User.find(session[:user_id]) if session[:user_id]
+  end
+
+  def logged_in?
+    !!current_user
+  end
+
+  def require_user
+    if !logged_in?
```
Comment (tlkiong) [inline app/controllers/application_controller.rb]:
```suggestion
    unless logged_in?
```

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -29,15 +30,34 @@ def update
...
+    session[:user_id] = nil
```
Comment (tlkiong) [inline app/controllers/users_controller.rb]: Question: what if `@user.destroy` has an error and the record is not destroyed? Your code here doesn't handle that case.

Your code here will show the notice and redirect but the record is not destroyed yet.

Fix it pls

**Review by tlkiong — CHANGES_REQUESTED** [review]

**Review by tlkiong — APPROVED** [review]

**Pull request merged**